node/cmd/spy/spy.go

```diff
@@ -136,8 +141,198 @@ func (s *spyServer) Publish(vaaBytes []byte) error {
+	s.subsAllVaaMu.Lock()
+	defer s.subsAllVaaMu.Unlock()
+
+	// this will try to unmarshal the byte array to a VAA, and then to a BatchVAA.
+	// the unmarshaling could populate one of the variable below. it's also possible
+	// a new type of vaaBytes comes through, so the variables below could remain empty.
+
+	var v *vaa.VAA
+	var b *vaa.BatchVAA
+
+	v, err := vaa.Unmarshal(vaaBytes)
+	// do nothing with the error, until we can try to unmarshal the bytes as a batch.
+	if err != nil {
+		// check if it is a batch
+
+		// it is not a VAA, try unmarshaling to a BatchVAA
+		b, err = vaa.UnmarshalBatch(vaaBytes)
+		if err != nil {
+			// it is not either type of VAA we know, nothing to do.
+			return err
+		}
+	}
+
+	// create the response(s) that will get sent out if this VAA satisfies a subscription.
+
+	// create the top-level response struct that is agnostic to the VAA type
+	var topRes *spyv1.SubscribeSignedVAAByTypeResponse
+
+	// if v has values, the VAA unmarshal was successful.
+	if len(v.Payload) > 0 {
+		// resData is the lowest level proto struct, it holds the byte data for whatever
+		// type of response it is (VAA in this case).
+		resData := &spyv1.SubscribeSignedVAAResponse{
+			VaaBytes: vaaBytes,
+		}
+
+		// resType defines what struct vaa will be retuned below, res of type SignedVaa
+		resType := &spyv1.SubscribeSignedVAAByTypeResponse_SignedVaa{
+			SignedVaa: resData,
+		}
+
+		// topRes is the highest level proto struct, the response to the subscription
+		topRes = &spyv1.SubscribeSignedVAAByTypeResponse{
+			// VaaType: &spyv1.SubscribeSignedVAAByTypeResponse_SignedVaa{
+			VaaType: resType,
+		}
+
+		// the proto is fully constructed ready to send.
+	}
+
+	// if b has vaules, the BatchVAA unmarshal was successful.
+	if len(b.Observations) > 0 {
+		// resData is the lowest level proto struct, it holds the byte data for whatever
+		// type of response it is (BatchVAA in this case).
+		resData := &spyv1.SubscribeSignedBatchVAAResponse{
+			BatchVaa: vaaBytes,
+		}
+
+		// resType defines what struct vaa will be retuned below, res of type SignedBatchVaa.
+		resType := &spyv1.SubscribeSignedVAAByTypeResponse_SignedBatchVaa{
+			SignedBatchVaa: resData,
+		}
+
+		// topRes is the highest level proto struct, the response to the subscription
+		topRes = &spyv1.SubscribeSignedVAAByTypeResponse{
+			// VaaType: &spyv1.SubscribeSignedVAAByTypeResponse_SignedVaa{
+			VaaType: resType,
+		}
+		// proto is fully constructed ready to send.
+	}
+
+	// loop through the subscriptions and send responses to everyone that wants this VAA
+	for _, sub := range s.subsAllVaa {
+		if len(sub.filters) == 0 {
+			// this subscription has no filters, send them the VAA.
+			sub.ch <- topRes
+		} else {
+			// this subscription has filters.
+
+			for i := range sub.filters {
+				// get the filter by index/GetFilter rather than from range,
+				// so as to not copy the mutex within the Filter.
```

Comment:
tl;dr - I fixed this by changing the `filters` struct to be an array of pointers, rather than an array of values.

`sub.filters` was `[]spyv1.FilterEntry`. When assigning the values in that range, ie `filterEntry` in the example below:
```golang
for _, filterEntry := range sub.filters {
  // ...
}
```

the values were being copied, which was throwing: 
> range var filterEntry copies lock:
> github.com/certusone/wormhole/node/pkg/proto/spy/v1.FilterEntry contains google.golang.org/protobuf/internal/impl.MessageState contains sync.Mutex `copylocks`


I changed `filters` to be references, rather than values, and updated this. Good question, thanks!